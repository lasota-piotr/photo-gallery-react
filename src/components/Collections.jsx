import React from 'react';
import {
  unsplashGetCollectionPhotos,
  unsplashListFeaturedCollections,
} from '../api/api';
import { LOAD_STATE } from '../constants/constants';

import CollectionsItem from './CollectionsItem';
import CollectionsLayout from './CollectionsLayout';
import PageTitle from './reusable/PageTitle';
import Wrapper from './reusable/Wrapper';
import PageLoading from './reusable/PageLoading';

class Collections extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      loadStateCollections: LOAD_STATE.LOADING,
      loadStateCollectionsPhotos: LOAD_STATE.LOADING,
      collections: [],
      collectionsPhotos: [],
    };
    this.fetchCollections = this.fetchCollections.bind(this);
  }

  componentDidMount() {
    this.fetchCollections();
  }

  fetchCollections() {
    unsplashListFeaturedCollections()
      .then(collections => {
        this.setState({
          collections,
          loadStateCollections: LOAD_STATE.SUCCESS,
        });
        return Promise.resolve(collections);
      })
      .then(collections =>
        Promise.all(
          collections.map(collection =>
            unsplashGetCollectionPhotos(collection.id, 1, 10, 'latest')
          )
        )
      )
      .then(collectionsPhotos => {
        this.setState({
          collectionsPhotos,
          loadStateCollectionsPhotos: LOAD_STATE.SUCCESS,
        });
      })
      .catch(() => {
        this.setState({
          loadStateCollections: LOAD_STATE.ERROR,
          loadStateCollectionsPhotos: LOAD_STATE.ERROR,
        });
      });
  }

  render() {
    const {
      loadStateCollections,
      collections,
      loadStateCollectionsPhotos,
      collectionsPhotos,
    } = this.state;
    return (
      <Wrapper>
        <PageTitle>Featured Collections</PageTitle>
        {loadStateCollections === LOAD_STATE.SUCCESS ? (
          <CollectionsLayout>
            {collections.map((collection, index) => (
              <CollectionsItem
                collectionInfo={collection}
                key={index}
                collectionPhotos={
                  collectionsPhotos.length > 0 ? collectionsPhotos[index] : []
                }
                isLoading={loadStateCollectionsPhotos === LOAD_STATE.LOADING}
                isError={loadStateCollectionsPhotos === LOAD_STATE.ERROR}
              />
            ))}
          </CollectionsLayout>
        ) : (
          <PageLoading />
        )}
      </Wrapper>
    );
  }
}

export default Collections;
