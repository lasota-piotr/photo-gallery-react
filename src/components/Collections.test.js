import React from 'react';
import renderer from 'react-test-renderer';
import Collections from './Collections';

global.fetch = jest.fn(() => new Promise(resolve => resolve()));

test('Renders Collections correctly', () => {
  const component = renderer.create(<Collections />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
