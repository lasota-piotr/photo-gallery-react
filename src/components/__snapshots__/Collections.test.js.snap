// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Collections correctly 1`] = `
<div
  className="sc-dnqmqq hrXOm"
>
  <h1
    className="sc-htoDjs eiveUc"
  >
    Featured Collections
  </h1>
  <div
    className="sc-iwsKbI jXxgDO"
  >
    <div
      className="sc-bdVaJa dbvdQh"
    >
      Loading
    </div>
  </div>
</div>
`;
